gdal/ogr/ogrsf_frmts/cad/gdalcaddataset.cpp

```diff
@@ -0,0 +1,463 @@
+                if(NULL == papszMetadata)
+                    SetMetadata(papszRasterMetadata, *papszDomainList);
+                else
+                    papszMetadata = CSLMerge(papszMetadata, papszRasterMetadata);    
+                papszDomainList++;
+            }
+        }
+    }
+
+    return TRUE;
+}
+
+OGRLayer *GDALCADDataset::GetLayer( int iLayer )
+{
+    if ( iLayer < 0 || iLayer >= nLayers )
+        return( NULL );
+    else
+        return( papoLayers[iLayer] );
+}
+
+int GDALCADDataset::TestCapability( const char * pszCap )
+{
+    if ( EQUAL(pszCap,ODsCCreateLayer) ||
+         EQUAL(pszCap,ODsCDeleteLayer) )
+    {
+         return FALSE;
+    }
+    else if( EQUAL(pszCap,ODsCCurveGeometries) )
+        return TRUE;
+    else if( EQUAL(pszCap,ODsCMeasuredGeometries) )
+        return TRUE;
+    return FALSE;
+}
+
+char** GDALCADDataset::GetFileList()
+{
+    size_t i, j;
+    char **papszFileList = GDALDataset::GetFileList();
+
+    /* duplicated papszFileList = CSLAddString( papszFileList, osCADFilename );*/
+    const char * pszPRJFilename = GetPrjFilePath();
+    if(NULL != pszPRJFilename)
+        papszFileList = CSLAddString( papszFileList, pszPRJFilename );
+
+    for(i = 0; i < poCADFile->GetLayersCount(); ++i)
+    {
+        CADLayer &oLayer = poCADFile->GetLayer( i );
+        for( j = 0; j < oLayer.getImageCount(); ++j )
+        {
+            CADImage* pImage = oLayer.getImage(j);
+            if(pImage)
+            {
+                CPLString osImgFilename = pImage->getFilePath(); 
+                if ( CPLCheckForFile ((char*)osImgFilename.c_str(), NULL) == TRUE)
+                    papszFileList = CSLAddString( papszFileList, osImgFilename );
+            }
+        }
+    }
+
+    if(NULL != poRasterDS)
+    {
+        papszFileList = CSLMerge(papszFileList, poRasterDS->GetFileList());
+    }
+    return papszFileList;
+}
+
+OGRSpatialReference *GDALCADDataset::GetSpatialReference()
+{
+    OGRSpatialReference *poSpatialRef = NULL;
+    if( poCADFile != NULL )
+    {
+        CPLString sESRISpatRef;
+        poSpatialRef = new OGRSpatialReference();
+        CADDictionary oNOD = poCADFile->GetNOD();
+        for( size_t i = 0; i < oNOD.getRecordsCount(); ++i )
+        {
+            if( !strcmp( oNOD.getRecord(i).first.c_str(), "ESRI_PRJ" ) )
+            {
+                CADXRecord * poXRecord = ( CADXRecord* ) oNOD.getRecord(i).second;
+                size_t dDataBegins = poXRecord->getRecordData().find("GEO");
+
+                std::string sESRISpatRefData( poXRecord->getRecordData().begin() + dDataBegins,
+                                              poXRecord->getRecordData().end() );
+                sESRISpatRef = sESRISpatRefData;
+                delete( poXRecord );
+        }
+    }
+
+        if( !sESRISpatRef.empty() )
+        {
+            char** papszPRJData = NULL;
+            papszPRJData = CSLAddString(papszPRJData, sESRISpatRef);
```

Comment:
Combine the above 2 lines:

`char** papszPRJData = CSLAddString(NULL, sESRISpatRef);`